RecoBTag/PixelCluster/plugins/PixelClusterTagInfoProducer.cc

```diff
@@ -0,0 +1,253 @@
+
+  // If no suitable Jet and PV is available, skip the event without opening pixel collection
+  if (collectionPVs->empty() || nJets <= 0) {
+    iEvent.put(std::move(pixelTagInfo));
+    return;
+  }
+
+  // Open Pixel Cluster collection
+  edm::Handle<edmNew::DetSetVector<SiPixelCluster> > collectionHandle;
+  iEvent.getByToken(m_pixelhit, collectionHandle);
+  const edmNew::DetSetVector<SiPixelCluster>& collectionClusters(*collectionHandle);
+
+  // Open Geometry
+  edm::ESHandle<TrackerGeometry> geom;
+  iSetup.get<TrackerDigiGeometryRecord>().get(geom);
+  const TrackerGeometry& theTracker(*geom);
+
+  // Retrieve tracker topology from geometry
+  edm::ESHandle<TrackerTopology> tTopoH;
+  iSetup.get<TrackerTopologyRcd>().get(tTopoH);
+  const TrackerTopology* tTopo = tTopoH.product();
+
+  std::vector<reco::PixelClusterProperties> clusters;
+
+  // Get vector of detunit ids, and fill a vector of PixelClusterProperties in the loop
+  for (auto const& detUnit : collectionClusters) {
+    if (detUnit.empty())
+      continue;
+    DetId detId = DetId(detUnit.detId());  // Get the Detid object
+    unsigned int detType = detId.det();    // det type, pixel = 1
+    if (detType != 1)
+      continue;                             // Consider only pixels
+    unsigned int subid = detId.subdetId();  // Subdetector type, pix barrel = 1, forward = 2
+    if (!(subid == PixelSubdetector::PixelBarrel || (m_addFPIX && subid == PixelSubdetector::PixelEndcap)))
+      continue;
+    unsigned int layer = tTopo->layer(detId);  // The layer index is in range 1-4 or 1-3
+    if (layer == 0 || layer > m_nLayers)
+      continue;
+
+    // Get the geom-detector
+    const auto* geomDet = theTracker.idToDet(detId);
+    const auto* topol = &geomDet->topology();
+
+    for (auto const& clUnit : detUnit) {
+      // get global position of the cluster
+      LocalPoint lp = topol->localPosition(MeasurementPoint(clUnit.x(), clUnit.y()));
+      GlobalPoint clustgp = geomDet->surface().toGlobal(lp);
+      if (m_minADC > 0 and clUnit.charge() < m_minADC)
+        continue;  // skip cluster if below threshold
+      reco::PixelClusterProperties cp = {clustgp.x(), clustgp.y(), clustgp.z(), clUnit.charge(), layer};
+      clusters.push_back(cp);
+    }
+  }
+
+  // Loop over jets and perform geometrical matching with pixel clusters
+  for (unsigned int j = 0, nj = collectionJets->size(); j < nj; j++) {
+    if (collectionJets->at(j).pt() < m_minJetPt)
+      continue;
+
+    edm::RefToBase<reco::Jet> jetRef = collectionJets->refAt(j);  // Get jet RefToBase
+
+    reco::PixelClusterData data(m_nLayers);
+    reco::PixelClusterTagInfo tagInfo;
+
+    for (auto const& cluster : clusters) {
+      GlobalPoint c3(cluster.x, cluster.y, cluster.z);
+      GlobalPoint v3(firstPV->x(), firstPV->y(), firstPV->z());
+      float dR = reco::deltaR(c3 - v3, jetRef->momentum());
+      float sC = m_hadronMass * 2. / (jetRef->pt());  // 2 mX / pT
```

Comment:
This doesn't need to be recomputed for every cluster: please move outside this loop